condor_history_to_prometheus.py

```diff
@@ -0,0 +1,135 @@
+#!/usr/bin/env python3
+import os, sys
+import glob
+from optparse import OptionParser
+import logging
+import htcondor, classad
+from condor_utils import *
+from condor_job_metrics import JobMetrics
+import datetime
+import time
+import prometheus_client
+from datetime import datetime
+import re
+
+utc_format = '%Y-%m-%dT%H:%M:%S'
+
+def generate_ads(entries):
+    for data in entries:
+        add_classads(data)
+        yield data
+
+def compose_ad_metrics(ad, metrics):
+    ''' Parse condor job classad and update metrics
+
+        Args:
+            ad (classad): an HTCondor job classad
+            metrics (JobMetrics): JobMetrics object 
+    '''
+    owner = ad['Owner']
+    site = ad['MATCH_EXP_JOBGLIDEIN_ResourceName']
+    schedd = ad['GlobalJobId'][0:ad['GlobalJobId'].find('#')]
+    walltimehrs = ad['walltimehrs']
+    device_name = ''
+    cpuhrs= ad['cpuhrs']
+
+    if ad['Requestgpus'] > 0:
+        device_name = ad['MachineAttrGPUs_DeviceName0']
+        gpuhrs = ad['gpuhrs']
+
+    # ignore this ad if walltimehrs is negative
+    if walltimehrs < 0:
+        return
+    
+    cpu_labels = [owner,site,schedd,device_name,'CPU']
+    gpu_labels = [owner,site,schedd,device_name,'GPU']
```

Comment:
Not entirely sure I understand. In case the label names change themselves or the values being provided for the labels change? 

So using something like `TypeDict` and making the label values types explict?

or setting up the labels with a dict like
```
labels = {owner: owner, site: site, schedd: schedd, GPUDeviceName: device_name, kind: None}
...
metrics.condor_job_walltime_hours.labels(**labels).inc(walltimehrs)
metrics.condor_job_resource_hours.labels(**labels, kind=).inc(cpuhrs)
```
(or both I suppose)

Reply:
The second is what I meant.  You could use various ways to achieve that (python dataclasses, namedtupe), but a dict is just fine.  Typing support is up to you.